### PR TITLE
Update `azurerm_linux_virtual_machine_scale_set` - Fix #8454

### LINKED ISSUE #8454: Default values for optional attributes of azurerm_linux_virtual_machine_scale_set resource forces replacement of resource
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform v0.12.25
+ provider.azurerm v2.26.0

### Affected Resource(s)

* `azurerm_linux_virtual_machine_scale_set`

### Terraform Configuration Files

Existing VMSS resource
```hcl
resource "azurerm_virtual_machine_scale_set" "manager" {
  name                = "manager"
  location            = azurerm_resource_group.rg.location
  resource_group_name = azurerm_resource_group.rg.name

  upgrade_policy_mode = "Manual"
  overprovision                      = false

  sku {
    name                             = "Standard_DS2_v2"
    tier                             = "Standard"
    capacity                         = 3
  }

  storage_profile_image_reference {
    id                              = var.image_id
  }

  storage_profile_os_disk {
    name                            = ""
    create_option                   = "FromImage"
    caching                         = "ReadWrite"
    managed_disk_type               = "Standard_LRS"
  }

  storage_profile_data_disk {
   lun                              = 0
   create_option                    = "FromImage"
   disk_size_gb                     = 128
  }

  storage_profile_data_disk {
   lun                              = 1
   create_option                    = "FromImage"
   disk_size_gb                     = 256
  }

  os_profile {
    computer_name_prefix            = "manager"
    admin_username                  = "localadmin"
  }

  os_profile_linux_config {
    disable_password_authentication = true
    ssh_keys {
      path                          = "/home/localadmin/.ssh/authorized_keys"
      key_data                      = var.admin_public_key 
    }
  }

  network_profile {
    name    = "primary"
    primary = true

    ip_configuration {
      name                                    = "manager-ipconfig"
      primary                                 = true
      subnet_id                               = azurerm_subnet.docker.id
    }
  }
}
```

Intended VMSS resource

```hcl
resource "azurerm_linux_virtual_machine_scale_set" "manager" {
  name                = "manager"
  location            = azurerm_resource_group.rg.location
  resource_group_name = azurerm_resource_group.rg.name
  sku                 = "Standard_DS2_v2"
  instances           = 3
  lifecycle {
    ignore_changes = [ instances ]
  }

  upgrade_mode = "Manual"

  overprovision                      = false

  source_image_id                    = var.manager_image_id

  os_disk {
    caching                         = "ReadWrite"
    storage_account_type            = "Standard_LRS"
  }


  data_disk {
      caching                   = "ReadWrite"
      create_option             = "FromImage"
      disk_size_gb              = 128
      lun                       = 0
      storage_account_type      = "Standard_LRS"
  }

  data_disk {
      caching                   = "ReadWrite"
      create_option             = "FromImage"
      disk_size_gb              = 256
      lun                       = 1
      storage_account_type      = "Standard_LRS"
  }

  computer_name_prefix            = "manager"

  admin_username                  = "localadmin"
  disable_password_authentication = true
  
  admin_ssh_key {
    username                        = "localadmin"
    public_key                      = var.admin_public_key
  }

  network_interface {
    name    = "primary"
    primary = true

    ip_configuration {
      name                                    = "manager-ipconfig"
      primary                                 = true
      subnet_id                               = azurerm_subnet.docker.id
    }
  }

}


```

### Expected Behavior

I am attempting to import an existing VMSS instance deployed using the `azurerm_virtual_machine_scale_set` into the new `azurem_linux_virtual_machine_scale_set` to provide an upgrade path of existing resources.

Although I am able to import the existing VMSS resource into the new azurerm_linux_virtual_machine_scale_set, when I later attempt to manage the imported resource via terraform plan/apply, terraform insists on replacing the resource.

The expected behaviour is that any new terraform configuration should be able to match the existing resource configuration and terraform should not replace the existing resource.

The state file of the original resource shows:

```
resource "azurerm_virtual_machine_scale_set" "manager" {
    automatic_os_upgrade   = false
    id                     = "/subscriptions/37d41301-c255-4f06-95b4-902a713aeb5e/resourceGroups/vnet/providers/Microsoft.Compute/virtualMachineScaleSets/manager"
    location               = "uksouth"
    name                   = "manager"
    overprovision          = false
    resource_group_name    = "vnet"
    single_placement_group = true
    upgrade_policy_mode    = "Manual"

 [...]
}
```

and 'az vmss list' shows:

```
{
[...]
    "upgradePolicy": {
      "automaticOsUpgradePolicy": {
        "disableAutomaticRollback": false,
        "enableAutomaticOsUpgrade": falsea
      },
      "mode": "Manual",
      "rollingUpgradePolicy": null
    },
    "virtualMachineProfile": {
       "priority": null,
    [...]
}
```

After terraform import, the state file shows:
```
{
[...]
   automatic_os_upgrade_policy {
        disable_automatic_rollback  = false
        enable_automatic_os_upgrade = false
    }
    upgrade_mode                                      = "Manual"
    max_bid_price                                     = -1
[...]
}
```

and `az vmss list` shows:

```
{
[...]
    "upgradePolicy": {
      "automaticOsUpgradePolicy": {
        "disableAutomaticRollback": false,
        "enableAutomaticOsUpgrade": false
      },
      "mode": "Manual",
      "rollingUpgradePolicy": null
    },
    "virtualMachineProfile": {
      "priority": null,
 [...]
 }
```

however `terraform plan`, then seeks to forces the 'priority' parameter to "Regular" even though it is unset in the state and azure configuration, which forces the scale set resource.

The expected behaviour is that the ability for the 'priority' parameter for this resource to be null within the azure configuration should be respected and there shouldn't be a default value if this parameter is not set in the terraform configuration.

In addition, `terraform import` also sets the `automatic_os_upgrade_policy` block in the state file, even though this block is optional when the `upgrade_mode` is set to `manual`.

The expected behaviour is that the optional nature of the `automatic_os_upgrade_policy` block should be respected and when a resource is imported where `upgrade_mode` is not set to `automatic`, this block should not appear in the resulting state file.

Also, `terraform import` also sets the `provision_vm_agent` to to be `true` in the state file of the imported resource.  There doesn't appear to be any azure configuration which correlates to this terraform option (that I could find).

The expected behaviour is that the lack of an existing azure configuration item correlating to this option should mean that there isn't any default value set if this parameter is not set in the terraform configuration.

Otherwise, these three issues prevent an upgrade path to the new `azurerm_linux_virtual_machine_scale_set` resource of any existing  deployed `azurerm_virtual_machine_scale_set` resources.

### Actual Behavior

```
  # azurerm_linux_virtual_machine_scale_set.manager must be replaced
-/+ resource "azurerm_linux_virtual_machine_scale_set" "manager" {
        admin_username                                    = "localadmin"
        computer_name_prefix                              = "manager"
        disable_password_authentication                   = true
        do_not_run_extensions_on_overprovisioned_machines = false
      ~ id                                                = "/subscriptions/37d41301-c255-4f06-95b4-902a713aeb5e/resourceGroups/vnet/providers/Microsoft.Compute/virtualMachineScaleSets/manager" -> (known after apply)
        instances                                         = 3
        location                                          = "uksouth"
        max_bid_price                                     = -1
        name                                              = "manager"
        overprovision                                     = false
      + priority                                          = "Regular" # forces replacement
        provision_vm_agent                                = true
        resource_group_name                               = "vnet"
        scale_in_policy                                   = "Default"
        single_placement_group                            = true
        sku                                               = "Standard_DS2_v2"
        source_image_id                                   = "/subscriptions/37d41301-c255-4f06-95b4-902a713aeb5e/resourceGroups/vnet/providers/Microsoft.Compute/images/manager"
      - tags                                              = {} -> null
      ~ unique_id                                         = "417a2673-5137-4a50-b9d9-a4572b05c3fa" -> (known after apply)
        upgrade_mode                                      = "Manual"
        zone_balance                                      = false
      - zones                                             = [] -> null

        admin_ssh_key {
            public_key = <<~EOT
                ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDOfyUiFZI0xn9C0HOoFGIn8c1MMESljLVPkuTLtytrbOTe/O62vO6YTeUbTYMS+ioHrjzCSjXI35OFrqsklp86njtCbdIHf8j1CTEsqrnyi+S9ExaF57oYZmPP33nkAjxHJrmPQQ+nCqKQUK6gPgMr3wHIGTaVeNznOkLBBRRwWQMNErQN8Lj+KdtJEjmpemVUUXTuNLtVsRSnTkQGUdKMYPoTXeZqfHH5PK8YoiqSPSq0JNcQaGYs/4iqwaYmpscceoi5VeakMTvvIkbLHUlYuGwjdMacTn4DlsN9ERfuh/JL1srp1IycRyCKI8fgossT1xgmYvPh3mp5WBnhrIwB boloughlin@roughscale
            EOT
            username   = "localadmin"
        }

      ~ automatic_instance_repair {
          ~ enabled      = false -> (known after apply)
          ~ grace_period = "PT30M" -> (known after apply)
        }

      - automatic_os_upgrade_policy { # forces replacement
          - disable_automatic_rollback  = false -> null
          - enable_automatic_os_upgrade = false -> null
        }

      ~ data_disk {
            caching                   = "ReadWrite"
            create_option             = "FromImage"
            disk_size_gb              = 128
            lun                       = 0
            storage_account_type      = "Standard_LRS"
            write_accelerator_enabled = false
        }
      ~ data_disk {
            caching                   = "ReadWrite"
            create_option             = "FromImage"
            disk_size_gb              = 256
            lun                       = 1
            storage_account_type      = "Standard_LRS"
            write_accelerator_enabled = false
        }

      + extension {
          + auto_upgrade_minor_version = (known after apply)
          + force_update_tag           = (known after apply)
          + name                       = (known after apply)
          + protected_settings         = (sensitive value)
          + provision_after_extensions = (known after apply)
          + publisher                  = (known after apply)
          + settings                   = (known after apply)
          + type                       = (known after apply)
          + type_handler_version       = (known after apply)
        }

      ~ network_interface {
          - dns_servers                   = [] -> null
            enable_accelerated_networking = false
            enable_ip_forwarding          = false
            name                          = "primary"
            primary                       = true

          ~ ip_configuration {
              - application_gateway_backend_address_pool_ids = [] -> null
              - application_security_group_ids               = [] -> null
              - load_balancer_backend_address_pool_ids       = [] -> null
              - load_balancer_inbound_nat_rules_ids          = [] -> null
                name                                         = "manager-ipconfig"
                primary                                      = true
                subnet_id                                    = "/subscriptions/37d41301-c255-4f06-95b4-902a713aeb5e/resourceGroups/vnet/providers/Microsoft.Network/virtualNetworks/vnet/subnets/docker"
                version                                      = "IPv4"
            }
        }

      ~ os_disk {
            caching                   = "ReadWrite"
          ~ disk_size_gb              = 32 -> (known after apply)
            storage_account_type      = "Standard_LRS"
            write_accelerator_enabled = false
        }

      + terminate_notification {
          + enabled = (known after apply)
          + timeout = (known after apply)
        }

      - timeouts {}
    }

Plan: 1 to add, 0 to change, 1 to destroy.
```
### Steps to Reproduce

1. `terraform apply (with codebase containing the azurerm_virtual_machine_scale_set resource)`
2. `terraform state rm azurerm_virtual_machine_scale_set.manager`
3. `RESOURCE_ID =$(az vmss list --query id)`
4. `terraform import azurerm_linux_virtual_machine_scale_set.manager $RESOURE_ID`
5. `terraform plan (with codebase containing the new azurerm_linux_virtual_machine_scale_set resource)`

### PR DESCRIPTION
**Pull request by ArcturusZhang**
Fixes https://github.com/terraform-providers/terraform-provider-azurerm/issues/8454

### CODE REVIEW

**Comment by katbyte** [comment]
@ArcturusZhang - isn't the problem here that we are setting a default? so shouldn't we just remove the default value?

File: azurerm/internal/services/compute/linux_virtual_machine_scale_set_resource.go [hunk 1]
```diff
@@ -896,7 +896,13 @@ func resourceArmLinuxVirtualMachineScaleSetRead(d *schema.ResourceData, meta int
 		d.Set("max_bid_price", maxBidPrice)
 
 		d.Set("eviction_policy", string(profile.EvictionPolicy))
-		d.Set("priority", string(profile.Priority))
+
+		// the service just return empty when this is not assigned when provisioned
```
Comment (tombuildsstuff) [inline azurerm/internal/services/compute/linux_virtual_machine_scale_set_resource.go:900]: fwiw this is a bug in the API that should be fixed - can we open a swagger issue for this?

**Comment by tombuildsstuff** [comment]
@katbyte in this instance the API is accepting an empty string and "Regular" as the same concept - but only returning this field if it's explicitly set, so this is an API bug where this field is conditionally nil, rather than always "Regular"

File: azurerm/internal/services/compute/linux_virtual_machine_scale_set_resource.go (hunk 1 above)
Comment (ArcturusZhang) [reply azurerm/internal/services/compute/linux_virtual_machine_scale_set_resource.go:900]: Upstream issue has opened: https://github.com/Azure/azure-rest-api-specs/issues/10971

**Pull request closed without merging**